File: DataStructures/SingleLinkedList/Node.cpp
#include "Node.h"
Node::Node(int val) {
	this->child = nullptr;
	this->value = val;
}
bool Node::Contains(int val) {
	if (val == this->value) return true;
	if (this->child != nullptr) {
		return this->child->Contains(val);
	}
	return false;
}
void Node::Remove(int val,Node*& tail) {
	if (this->child->value == val) {
		Node* childNode = this->child;
		this->child = this->child->child;
		delete childNode;
		childNode = nullptr;

		//Assign new tail if child node is the new last node 
		if (this->child == nullptr) {
			tail = this;
		}
	}
	else {
		this->child->Remove(val,tail);
	}
}
void Node::CleanUp() {
	if (this->child != nullptr) { 
		this->child->CleanUp(); 
	}
	delete this;
}

File: src/linked/Node.ts
import {INode} from "./INode";
export abstract class Node implements INode{
    child?:INode;
    value:number;
    constructor(val:number,childNode?:INode) {
        this.value = val;
        this.child = childNode;
    }
    abstract insert(val: number): INode|null;
    search(val: number): INode | null {
        if(this.value === val) return this;
        if(this.child){
            return this.child.search(val);
        }
        return null;
    }
    abstract remove(val: number): INode|null;
}